File: src/routes/jobs.js
const express = require('express');
const {
  listJobs,
  getJobById,
  updateJobStatus,
  canRunJob,
  addJobLog,
  getJobLogs,
} = require('../services/jobs');
const { runJob } = require('../services/executor');
const { getQueueSnapshot, scheduleJob } = require('../services/scheduler');

const router = express.Router();

router.get('/jobs', (req, res) => {
  res.json({
    jobs: listJobs(),
  });
});

router.get('/queue', (req, res) => {
  res.json(getQueueSnapshot());
});

router.get('/jobs/:id', (req, res) => {
  const job = getJobById(req.params.id);

  if (!job) {
    return res.status(404).json({
      error: 'Job not found',
    });
  }

  return res.json(job);
});

router.get('/jobs/:id/logs', (req, res) => {
  const result = getJobLogs(req.params.id);

  if (!result.ok) {
    return res.status(404).json({
      error: result.reason,
    });
  }

  return res.json({
    logs: result.logs,
  });
});

router.patch('/jobs/:id/status', (req, res) => {
  const { status } = req.body;
  const result = updateJobStatus(req.params.id, status);

  if (!result.ok) {
    const statusCode = result.reason === 'Job not found' ? 404 : 400;

    return res.status(statusCode).json({
      error: result.reason,
    });
  }

  return res.json({
    message: 'Job status updated',
    job: result.job,
  });
});

router.post('/jobs/:id/run', async (req, res) => {
  const runnable = canRunJob(req.params.id);

  if (!runnable.ok) {
    const statusCode = runnable.reason === 'Job not found' ? 404 : 400;

    return res.status(statusCode).json({
      error: runnable.reason,
    });
  }

  const result = await runJob(req.params.id);

  if (!result.ok) {
    const statusCode = result.reason === 'Job not found' ? 404 : 400;

    return res.status(statusCode).json({
      error: result.reason,
    });
  }

  return res.json({
    message: 'Job run completed',
    job: result.job,
  });
});

router.post('/jobs/:id/schedule', (req, res) => {
  const result = scheduleJob(req.params.id);

  if (!result.ok) {
    const statusCode = result.reason === 'Job not found' ? 404 : 400;

    return res.status(statusCode).json({
      error: result.reason,
    });
  }

  return res.status(202).json({
    message: 'Job scheduled',
    job: result.job,
  });
});

router.post('/jobs/:id/logs', (req, res) => {
  const { level, message } = req.body;
  const result = addJobLog(req.params.id, { level, message });

  if (!result.ok) {
    const statusCode = result.reason === 'Job not found' ? 404 : 400;

    return res.status(statusCode).json({
      error: result.reason,
    });
  }

  return res.status(201).json({
    message: 'Job log added',
    log: result.entry,
  });
});

module.exports = router;

File: app/routers/jobs.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException

from ..schemas import JobLogCreate, JobStatusUpdate
from ..services.jobs import (
    add_job_log,
    can_run_job,
    get_job_by_id,
    get_job_logs,
    list_jobs,
    update_job_status,
)
from ..services.scheduler import scheduler

router = APIRouter(prefix="/jobs", tags=["jobs"])


@router.get("")
def jobs_list() -> dict:
    return {"jobs": list_jobs()}


@router.get("/{job_id}")
def jobs_get(job_id: str):
    job = get_job_by_id(job_id)
    if not job:
        raise HTTPException(status_code=404, detail="Job not found")
    return job


@router.get("/{job_id}/logs")
def jobs_logs(job_id: str) -> dict:
    result = get_job_logs(job_id)
    if not result["ok"]:
        raise HTTPException(status_code=404, detail=result["reason"])
    return {"logs": result["logs"]}


@router.patch("/{job_id}/status")
def jobs_update_status(job_id: str, body: JobStatusUpdate) -> dict:
    result = update_job_status(job_id, body.status)
    if not result["ok"]:
        raise HTTPException(
            status_code=404 if result["reason"] == "Job not found" else 400,
            detail=result["reason"],
        )
    return {"message": "Job status updated", "job": result["job"]}


@router.post("/{job_id}/schedule", status_code=202)
def jobs_schedule(job_id: str) -> dict:
    result = scheduler.schedule_job(job_id)
    if not result["ok"]:
        raise HTTPException(
            status_code=404 if result["reason"] == "Job not found" else 400,
            detail=result["reason"],
        )
    return {"message": "Job scheduled", "job": result["job"]}


@router.post("/{job_id}/run")
def jobs_run(job_id: str) -> dict:
    from ..services.executor import run_job

    result = can_run_job(job_id)
    if not result["ok"]:
        raise HTTPException(
            status_code=404 if result["reason"] == "Job not found" else 400,
            detail=result["reason"],
        )

    execution_result = run_job(job_id)
    if not execution_result["ok"]:
        raise HTTPException(
            status_code=404 if execution_result["reason"] == "Job not found" else 400,
            detail=execution_result["reason"],
        )

    return {"message": "Job run completed", "job": execution_result["job"]}


@router.post("/{job_id}/logs", status_code=201)
def jobs_add_log(job_id: str, body: JobLogCreate) -> dict:
    result = add_job_log(job_id, level=body.level, message=body.message)
    if not result["ok"]:
        raise HTTPException(
            status_code=404 if result["reason"] == "Job not found" else 400,
            detail=result["reason"],
        )
    return {"message": "Job log added", "log": result["entry"]}
